File: sidecar/app/session/controllers/session_controller.py
"""HTTP handlers for session CRUD operations."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, status

from sidecar.app.session.controllers.responses.session_response import SessionResponse
from sidecar.app.session.controllers.validations.create_session_validation import CreateSessionValidation
from sidecar.app.session.controllers.validations.update_session_validation import UpdateSessionValidation
from sidecar.app.session.domain.abstractions.session_service import SessionService
from sidecar.app.session.domain.dtos.create_session_dto import CreateSessionDto
from sidecar.app.session.domain.dtos.update_session_dto import UpdateSessionDto
from sidecar.http.state import get_app_state
from sidecar.shared.http_responses import HttpSuccessResponse, build_success_response

router = APIRouter(prefix="/sessions", tags=["sessions"])


@router.post("", response_model=HttpSuccessResponse[SessionResponse])
async def create_session(request_body: CreateSessionValidation) -> HttpSuccessResponse[SessionResponse]:
    """Create a new session."""
    session_service = _get_session_service()
    session = await session_service.create(CreateSessionDto(session_name=request_body.session_name))
    return build_success_response(session)


@router.get("", response_model=HttpSuccessResponse[list[SessionResponse]])
async def get_sessions() -> HttpSuccessResponse[list[SessionResponse]]:
    """Return all sessions."""
    session_service = _get_session_service()
    sessions = await session_service.get()
    return build_success_response(sessions)


@router.get("/{session_id}", response_model=HttpSuccessResponse[SessionResponse])
async def get_session_by_id(session_id: str) -> HttpSuccessResponse[SessionResponse]:
    """Return a single session."""
    session_service = _get_session_service()
    session = await session_service.get_one_by_id(session_id)
    return build_success_response(session)


@router.patch("/{session_id}", response_model=HttpSuccessResponse[SessionResponse])
async def update_session_by_id(
    session_id: str,
    request_body: UpdateSessionValidation,
) -> HttpSuccessResponse[SessionResponse]:
    """Update a session."""
    session_service = _get_session_service()
    session = await session_service.update_by_id(
        session_id,
        UpdateSessionDto(session_name=request_body.session_name),
    )
    return build_success_response(session)


@router.patch("/{session_id}/end", response_model=HttpSuccessResponse[SessionResponse])
async def end_session_by_id(session_id: str) -> HttpSuccessResponse[SessionResponse]:
    """End a session."""
    session_service = _get_session_service()
    session = await session_service.end_by_id(session_id)
    return build_success_response(session)


@router.delete("/{session_id}", status_code=status.HTTP_200_OK, response_model=HttpSuccessResponse[None])
async def delete_session_by_id(session_id: str) -> HttpSuccessResponse[None]:
    """Delete a session."""
    session_service = _get_session_service()
    await session_service.delete_by_id(session_id)
    return build_success_response(None)


def _get_session_service() -> SessionService:
    state = get_app_state()
    session_service = state.session_service
    if session_service is None:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Session service is not ready.")
    return session_service

File: sidecar/src/app/session/controllers/session-controller.ts
import type { NextFunction, Request, Response } from 'express';
import type { SessionService } from '../domain/abstractions/session-service';
import { sendSuccessResponse } from '../../../shared/http-responses';
import { getStringParam } from '../../../shared/utils/http';

export class SessionController {
  constructor(private readonly sessionService: SessionService) {}

  createSession = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const session = await this.sessionService.create(req.body);
      sendSuccessResponse(res, session, 'Session created successfully', 201);
    } catch (error) {
      next(error);
    }
  };

  getSessions = async (_req: Request, res: Response, next: NextFunction) => {
    try {
      const sessions = await this.sessionService.get();
      sendSuccessResponse(res, sessions, 'Sessions fetched successfully');
    } catch (error) {
      next(error);
    }
  };

  getSessionById = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const sessionId = getStringParam(req.params.sessionId);
      const session = await this.sessionService.getOneById(sessionId);
      sendSuccessResponse(res, session, 'Session fetched successfully');
    } catch (error) {
      next(error);
    }
  };

  updateSessionById = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const sessionId = getStringParam(req.params.sessionId);
      const session = await this.sessionService.updateById(sessionId, req.body);
      sendSuccessResponse(res, session, 'Session updated successfully');
    } catch (error) {
      next(error);
    }
  };

  endSessionById = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const sessionId = getStringParam(req.params.sessionId);
      const session = await this.sessionService.endById(sessionId);
      sendSuccessResponse(res, session, 'Session ended successfully');
    } catch (error) {
      next(error);
    }
  };

  deleteSessionById = async (req: Request, res: Response, next: NextFunction) => {
    try {
      const sessionId = getStringParam(req.params.sessionId);
      await this.sessionService.deleteById(sessionId);
      sendSuccessResponse(res, null, 'Session deleted successfully');
    } catch (error) {
      next(error);
    }
  };
}
